Library/Homebrew/cmd/update-report.rb

```diff
@@ -167,16 +170,113 @@ def migrate_legacy_cache_if_necessary
     end
   end
 
-  def link_completions_and_docs
-    return if HOMEBREW_PREFIX.to_s == HOMEBREW_REPOSITORY.to_s
+  def migrate_legacy_repository_if_necessary
+    return unless HOMEBREW_PREFIX.to_s == "/usr/local"
+    return unless HOMEBREW_REPOSITORY.to_s == "/usr/local"
+    return unless ARGV.homebrew_developer?
+
+    unless HOMEBREW_PREFIX.writable_real?
+      onoe <<-EOS.undent
+        #{HOMEBREW_PREFIX} is not writable.
+
+        You should change the ownership and permissions of #{HOMEBREW_PREFIX}
+        temporarily back to your user account so we can complete the Homebrew
+        repository migration:
+          sudo chown -R $(whoami) #{HOMEBREW_PREFIX}
+      EOS
+    end
+
+    Keg::ALL_TOP_LEVEL_DIRECTORIES.each do |dir|
+      FileUtils.mkdir_p "#{HOMEBREW_PREFIX}/#{dir}"
+    end
+
+    new_homebrew_repository = Pathname.new "/usr/local/Homebrew"
+    if new_homebrew_repository.exist? && !new_homebrew_repository.children.empty?
+      onoe <<-EOS.undent
+        #{new_homebrew_repository} already exists.
+        Please remove it manually or uninstall and reinstall Homebrew into a new
+        location as the migration cannot be done automatically.
+      EOS
+    end
+    new_homebrew_repository.mkpath
+
+    repo_files = HOMEBREW_REPOSITORY.cd do
+      Utils.popen_read("git ls-files").lines.map(&:chomp)
+    end
+
+    FileUtils.cp_r "#{HOMEBREW_REPOSITORY}/.git", "#{new_homebrew_repository}/.git"
+    new_homebrew_repository.cd do
+      safe_system "git", "checkout", "."
+    end
+
+    if (HOMEBREW_REPOSITORY/"Library/Taps").exist?
+      FileUtils.cp_r "#{HOMEBREW_REPOSITORY}/Library/Taps", "#{new_homebrew_repository}/Library/Taps"
+    end
+
+    if (HOMEBREW_REPOSITORY/"Library/LinkedKegs").exist?
+      FileUtils.mkdir_p "#{new_homebrew_repository}/Library/LinkedKegs"
+      (HOMEBREW_REPOSITORY/"Library/LinkedKegs").children.each do |link|
+        src = link.realpath
+        dst = new_homebrew_repository/"Library/LinkedKegs/#{link.basename}"
+        FileUtils.ln_sf(dst.relative_path_from(src.parent), src)
+      end
+    end
+
+    src = Pathname.new("#{new_homebrew_repository}/bin/brew")
+    dst = Pathname.new("#{HOMEBREW_PREFIX}/bin/brew")
+    begin
+      FileUtils.ln_sf(dst.relative_path_from(src.parent), src)
+    rescue Errno::EACCES
+      onoe <<-EOS.undent
+        Could not create symlink at #{dst}!
+        Please do this manually with:
+          sudo ln -sf #{src} #{dst}
+          sudo chown $(whoami) #{dst}
+      EOS
+    end
+
+    unremovable_paths = []
+    remove_dirs = [".git", "share/doc/homebrew/images", "share/doc/homebrew"]
+    (repo_files + remove_dirs).each do |file|
+      path = Pathname.new "#{HOMEBREW_REPOSITORY}/#{file}"
+      begin
+        FileUtils.rm_rf path
+      rescue Errno::EACCES
+        unremovable_paths << path
+      end
+      path.parent.rmdir_if_possible
```

Comment:
I think we need a `rmdir -p` here, or `Library` for instance won't be removed even if empty, because when we got to its immediate children there was still this `Library/Homebrew` in the way.